contracts/Marketplace.sol

```diff
@@ -408,6 +408,14 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
     _;
   }
 
+  modifier requestNotFinished(RequestId requestId) {
+    require(
+      _requestContexts[requestId].state != RequestState.Finished,
+      "Request is cancelled or finished"
+    );
```

Comment:
Nice! But, I don't think this is enough. You should use instead `requestState(requestId)` which checks dynamically on the request's state, which is not always "propagated" into the `requestContext.state`variable. For example `Cancelled` state is saved to `requestContext.state` only once the Client calls `withdrawFunds`. Also `Finished` != `Cancelled`... 

I am curious, if the prover is passing with this fix, maybe there is some over constraining happening in your rules? Because I would have expected that it should find counter-examples for this, no?

Reply:
Very good catch. I remember actually using `requestState()` while debugging something and then I rolled back to this.

Will verify with `requestState()` again and update this commit. Thank you for pointing this out!

Reply:
@AuHau passes with `requestState()` 

https://prover.certora.com/output/6199/64289db2b3044f21937c1955f1bf695a?anonymousKey=35a74f14ec6be98fe9fab29f03187cfd5c810178